toplevel/identity/entity.go

```diff
@@ -0,0 +1,642 @@
+	"github.com/app-sre/vault-manager/toplevel"
+	"github.com/hashicorp/go-version"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+type config struct{}
+
+var _ toplevel.Configuration = config{}
+
+type user struct {
+	Name        string `yaml:"name"`
+	OrgUsername string `yaml:"org_username"`
+	Roles       []role `yaml:"roles"`
+}
+
+type role struct {
+	Name        string           `yaml:"name"`
+	Permissions []oidcPermission `yaml:"oidc_permissions"`
+}
+
+type oidcPermission struct {
+	Name        string        `yaml:"name"`
+	Description string        `yaml:"description"`
+	Service     string        `yaml:"service"`
+	Policies    []vaultPolicy `yaml:"vault_policies"`
+}
+
+type vaultPolicy struct {
+	Name string `yaml:"name"`
+}
+
+type entity struct {
+	Name     string
+	Id       string
+	Type     string
+	Metadata interface{}
+	Policies []string
+	Aliases  []entityAlias
+}
+
+type entityAlias struct {
+	Name           string
+	Id             string
+	Type           string
+	AuthType       string
+	AccessorId     string
+	CustomMetadata interface{}
+}
+
+var _ vault.Item = entity{}
+
+var _ vault.Item = entityAlias{}
+
+func (e entity) Key() string {
+	return e.Name
+}
+
+func (e entity) KeyForType() string {
+	return e.Type
+}
+
+func (e entity) KeyForDescription() string {
+	return fmt.Sprintf("%v", e.Metadata)
+}
+
+func (e entity) Equals(i interface{}) bool {
+	entry, ok := i.(entity)
+	if !ok {
+		return false
+	}
+	return e.Name == entry.Name &&
+		reflect.DeepEqual(e.Metadata, entry.Metadata) &&
+		reflect.DeepEqual(e.Policies, entry.Policies)
+}
+
+func (e entity) CreateOrUpdate(action string) {
+	path := filepath.Join("identity", e.Type, "name", e.Name)
+	config := map[string]interface{}{
+		"policies": e.Policies,
+		"metadata": e.Metadata,
+	}
+	vault.WriteSecret(path, config)
+	log.WithField("path", path).WithField("type", e.Type).Info(
+		fmt.Sprintf("[Vault Identity] entity successfully %s", action))
+}
+
+func (e entity) Delete() {
+	path := filepath.Join("identity", e.Type, "name", e.Name)
+	vault.DeleteSecret(path)
+	log.WithField("path", path).WithField("type", e.Type).Info(
+		"[Vault Identity] entity successfully deleted")
+}
+
+func (e entityAlias) Key() string {
+	return e.Name
+}
+
+func (e entityAlias) KeyForType() string {
+	return e.Type
+}
+
+func (e entityAlias) KeyForDescription() string {
+	return fmt.Sprintf("%v", e.CustomMetadata)
+}
+
+func (e entityAlias) Equals(i interface{}) bool {
+	entry, ok := i.(entityAlias)
+	if !ok {
+		return false
+	}
+	return e.Name == entry.Name &&
+		e.AuthType == entry.AuthType &&
+		reflect.DeepEqual(e.CustomMetadata, entry.CustomMetadata)
+}
+
+func (ea entityAlias) Create(entityId string, customMetadataSupported bool) {
+	path := filepath.Join("identity", ea.Type)
+	config := map[string]interface{}{
+		"name":           ea.Name,
+		"canonical_id":   entityId,
+		"mount_accessor": ea.AccessorId,
+	}
+	if customMetadataSupported {
```

Comment:
is there a use case for customMetadata?

Reply:
Unsure. I opted to write the logic to include it as it seemed easier to remove after the fact than add it in later. Maybe for trivial info like which team someone is apart of etc. Can easily rip out if need be.